src/impute_srp_model.cpp

```diff
@@ -0,0 +1,201 @@
+    t, shape_progression, scale_progression, 1, 0// lower = TRUE, log = FALSE
+  );
+  // use formula of Bayes to calculate conditional probability of response
+  const double pr_response = p * pr_survival_response / (
+    p * pr_survival_response + (1 - p) * pr_survival_progression
+  );
+  return pr_response;
+}
+
+// [[Rcpp::export]]
+DataFrame impute_srp_model(
+    DataFrame df,
+    NumericMatrix p,
+    NumericVector shape_vec,
+    NumericVector scale_vec,
+    NumericVector visit_spacing,
+    int n_sim, // number of resamples
+    int n_groups
+) {
+  // extract columns from df
+  IntegerVector subject_id = df["subject_id"];
+  IntegerVector group_id = df["group_id"];
+  CharacterVector from = df["from"];
+  CharacterVector to = df["to"];
+  NumericVector t_min = df["t_min"];
+  NumericVector t_max = df["t_max"];
+  NumericVector t_sot = df["t_sot"];
+  // store shape/scale as multidimensional arrays
+  IntegerVector dims = {n_sim, n_groups, 3};
+  Function array_ind("arrayInd");
+  double shape[n_sim][n_groups][3];
+  double scale[n_sim][n_groups][3];
+  for (int i = 0; i < shape_vec.length(); i++) {
+    IntegerVector idx = array_ind(i + 1, dims);
+    shape[idx[0] - 1][idx[1] - 1][idx[2] - 1] = shape_vec[i];
+    scale[idx[0] - 1][idx[1] - 1][idx[2] - 1] = scale_vec[i];
+  }
+  // define output vectors of length 0
+  IntegerVector subject_id_out (0);
+  IntegerVector group_id_out (0);
+  CharacterVector from_out (0);
+  CharacterVector to_out (0);
+  NumericVector t_min_out (0);
+  NumericVector t_max_out (0);
+  NumericVector t_sot_out (0);
+  IntegerVector iter_out (0);
+  // iterate over number of samples to draw
+  for (int iter = 0; iter < n_sim; iter++) {
+    // iterate over rows in data
+    for (int i = 0; i < df.nrows(); i++) {
+      int g = group_id[i] - 1; // -1 due to 0 based indexing in C
+      bool target_state_known = !CharacterVector::is_na(to[i]);
+      if (target_state_known) {
+        // nothing to sample - target state is observed, copy for each iteration
+        subject_id_out.push_back(subject_id(i));
+        group_id_out.push_back(group_id(i));
+        from_out.push_back(from(i));
+        to_out.push_back(to(i));
+        t_min_out.push_back(t_min(i));
+        t_max_out.push_back(t_max(i));
+        t_sot_out.push_back(t_sot(i));
+        iter_out.push_back(iter + 1);
+      } else {
+        if (from(i) == "stable") {
+          // first sample whether response or progression happens
+          double pr_response = conditional_response_probability_srp(
+              t_min(i) - t_sot(i), // (min) time since start of treatment
+              p(iter, g), // unconditional response probability
+              // transition index -1 due to 0 based indexing in C
+              shape[iter][g][0], shape[iter][g][1],
+              scale[iter][g][0], scale[iter][g][1]
+            );
+          bool response = R::rbinom(1, pr_response);
+          if (response) {
```

Comment:
this seems to be where the current discrepancy comes from @AKktk almost certain it is an offset issue

Reply:
checked the if(response){ } block